File: src/sync/folders.py
"""List IMAP folders for a configured account.

Usage:
  corrkit list-folders <account-name>
  corrkit list-folders              # lists available account names
"""

import argparse
import ssl as ssl_mod

from imapclient import IMAPClient

from accounts import load_accounts_or_env, resolve_password


def main() -> None:
    parser = argparse.ArgumentParser(description="List IMAP folders for an account")
    parser.add_argument("account", nargs="?", help="Account name from accounts.toml")
    args = parser.parse_args()

    accounts = load_accounts_or_env()

    if not args.account:
        print("Available accounts:")
        for name, acct in accounts.items():
            print(f"  {name:<20} {acct.user}")
        return

    if args.account not in accounts:
        raise SystemExit(
            f"Unknown account: {args.account}\n"
            f"Available: {', '.join(accounts.keys())}"
        )

    acct = accounts[args.account]
    password = resolve_password(acct)

    ssl_context = None
    if acct.imap_starttls or acct.imap_host in ("127.0.0.1", "localhost"):
        ssl_context = ssl_mod.create_default_context()
        ssl_context.check_hostname = False
        ssl_context.verify_mode = ssl_mod.CERT_NONE

    use_ssl = not acct.imap_starttls
    print(f"Connecting to {acct.imap_host}:{acct.imap_port} as {acct.user}\n")

    with IMAPClient(
        acct.imap_host, port=acct.imap_port,
        ssl=use_ssl, ssl_context=ssl_context,
    ) as imap:
        if acct.imap_starttls:
            imap.starttls(ssl_context=ssl_context)
        imap.login(acct.user, password)
        for flags, _delim, folder_name in imap.list_folders():
            flag_str = ", ".join(
                f.decode() if isinstance(f, bytes) else str(f) for f in flags
            )
            print(f"  {folder_name:<40} [{flag_str}]")


if __name__ == "__main__":
    main()

File: src/sync/folders.rs
//! List IMAP folders for a configured account.

use anyhow::Result;

use crate::accounts::{load_accounts, resolve_password};

pub fn run(account: Option<&str>) -> Result<()> {
    let accounts = load_accounts(None)?;

    let Some(account_name) = account else {
        println!("Available accounts:");
        for (name, acct) in &accounts {
            println!("  {:<20} {}", name, acct.user);
        }
        return Ok(());
    };

    let acct = accounts.get(account_name).ok_or_else(|| {
        anyhow::anyhow!(
            "Unknown account: {}\nAvailable: {}",
            account_name,
            accounts.keys().cloned().collect::<Vec<_>>().join(", ")
        )
    })?;
    let password = resolve_password(acct)?;

    let mut tls_builder = native_tls::TlsConnector::builder();
    if acct.imap_starttls || acct.imap_host == "127.0.0.1" || acct.imap_host == "localhost" {
        tls_builder.danger_accept_invalid_certs(true);
        tls_builder.danger_accept_invalid_hostnames(true);
    }
    let tls = tls_builder.build()?;

    println!(
        "Connecting to {}:{} as {}\n",
        acct.imap_host, acct.imap_port, acct.user
    );

    let client = if acct.imap_starttls {
        imap::connect_starttls(
            (acct.imap_host.as_str(), acct.imap_port),
            &acct.imap_host,
            &tls,
        )?
    } else {
        imap::connect(
            (acct.imap_host.as_str(), acct.imap_port),
            &acct.imap_host,
            &tls,
        )?
    };

    let mut session = client.login(&acct.user, &password).map_err(|e| e.0)?;
    let folders = session.list(None, Some("*"))?;

    for folder in folders.iter() {
        let flags: Vec<String> = folder
            .attributes()
            .iter()
            .map(|a| format!("{:?}", a))
            .collect();
        println!("  {:<40} [{}]", folder.name(), flags.join(", "));
    }

    session.logout()?;
    Ok(())
}
